hudi-utilities/src/main/java/org/apache/hudi/utilities/MORRestoreTool.java

```diff
@@ -0,0 +1,329 @@
+      LOG.error("Fail to do hoodie metadata table validation for " + validator.cfg, throwable);
+    } finally {
+      jsc.stop();
+    }
+  }
+
+  public boolean run() {
+    boolean result = false;
+    try {
+      LOG.info(cfg);
+      LOG.info(" ****** Triggering restore to " + cfg.commitTime + " ******");
+      result = doRestore();
+    } catch (Exception e) {
+      throw new HoodieException("Unable to do hoodie metadata table validation in " + cfg.basePath, e);
+    } finally {
+      return result;
+    }
+  }
+
+  public boolean doRestore() {
+    boolean finalResult = true;
+    String basePath = metaClient.getBasePath();
+    HoodieSparkEngineContext engineContext = new HoodieSparkEngineContext(jsc);
+    HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().enable(false).build();
+    List<String> partitions = FSUtils.getAllPartitionPaths(engineContext, metadataConfig, basePath);
+    Map<String, List<String>> filesToDelete = new HashMap<>();
+
+    HoodieTableFileSystemView fileSystemView = FileSystemViewManager.createInMemoryFileSystemView(engineContext, metaClient, metadataConfig);
+    partitions.forEach(pPath -> {
+      LOG.info("Processing partition " + pPath);
+      filesToDelete.put(pPath, new ArrayList<>());
+      List<FileSlice> fileSlices = fileSystemView.getAllFileSlices(pPath).collect(Collectors.toList());
+      for (FileSlice fileSlice : fileSlices) {
+        LOG.info("File slice commit time " + fileSlice.getBaseInstantTime());
+        if (HoodieTimeline.compareTimestamps(fileSlice.getBaseInstantTime(), HoodieTimeline.GREATER_THAN, cfg.commitTime)) {
+          LOG.info("Deleting entire file slice ");
+          if (fileSlice.getBaseFile().isPresent()) {
+            LOG.info("Base file to delete " + fileSlice.getBaseFile().get().getPath());
+            filesToDelete.get(pPath).add(fileSlice.getBaseFile().get().getPath());
+          }
+          fileSlice.getLogFiles().forEach(logFile -> {
+            LOG.info("   log file to delete " + logFile.getPath().toString());
+            filesToDelete.get(pPath).add(logFile.getPath().toString());
+          });
+        } else if (HoodieTimeline.compareTimestamps(fileSlice.getBaseInstantTime(), HoodieTimeline.EQUALS, cfg.commitTime)) {
+          LOG.info("Deleting all log files except base file");
+          if (fileSlice.getBaseFile().isPresent()) {
+            LOG.info("Not deleting Base file " + fileSlice.getBaseFile().get().getPath());
+          }
+          fileSlice.getLogFiles().forEach(logFile -> {
+            LOG.info("   log file to delete " + logFile.getPath().toString());
+            filesToDelete.get(pPath).add(logFile.getPath().toString());
+          });
+          LOG.info("Not procssing remaining file slices");
+          break;
+        } else {
+          // we need to collect only partial list of log files to be deleted
+          /*TableSchemaResolver tableSchemaResolver = new TableSchemaResolver(metaClient);
```

Comment:
It is feasible to support restoring to any delta commit. For now, have not tested this part yet.